runtime/core/backend/inverse_text_normalizer.cc

```diff
@@ -0,0 +1,238 @@
+  }
+
+  if (output_mode == "byte") {
+    type_ = BYTE;
+  } else if (output_mode == "utf8") {
+    type_ = UTF8;
+  } else {
+    type_ = SYMBOL;
+    output_symtab_ = std::shared_ptr<fst::SymbolTable>(
+        fst::SymbolTable::ReadText(output_mode));
+    if (!output_symtab_) LOG(FATAL) << "Invalid mode or symbol table path.";
+  }
+}
+
+const std::string InverseTextNormalizer::ProcessInput(
+    const std::string& input) {
+  fst::StdVectorFst input_fst, output_fst;
+  if (!(compiler_->operator()(input, &input_fst))) {
+    return "Unable to parse input string.";
+  }
+  FormatFst(&input_fst);
+  string return_val = "";
+  // Set symbols for the input, if appropriate
+  if (byte_symtab_ && type_ == BYTE) {
+    input_fst.SetInputSymbols(byte_symtab_.get());
+    input_fst.SetOutputSymbols(byte_symtab_.get());
+  } else if (utf8_symtab_ && type_ == UTF8) {
+    input_fst.SetInputSymbols(utf8_symtab_.get());
+    input_fst.SetOutputSymbols(utf8_symtab_.get());
+  } else if (input_symtab_ && type_ == SYMBOL) {
+    input_fst.SetInputSymbols(input_symtab_.get());
+    input_fst.SetOutputSymbols(input_symtab_.get());
+  }
+
+  bool succeeded = true;
+  for (size_t i = 0; i < rules_.size(); ++i) {
+    thrax::RuleTriple triple(rules_[i]);
+    if (grm_.Rewrite(triple.main_rule, input_fst, &output_fst,
+                     triple.pdt_parens_rule, triple.mpdt_assignments_rule)) {
+      input_fst = output_fst;
+    } else {
+      succeeded = false;
+      break;
+    }
+  }
+
+  std::vector<std::pair<string, float> > strings;
+  std::set<string> seen;
+  if (succeeded && FstToStrings(output_fst, 1, &strings)) {
+    std::vector<std::pair<string, float> >::iterator itr = strings.begin();
+    for (; itr != strings.end(); ++itr) {
+      std::set<string>::iterator sx = seen.find(itr->first);
+      if (sx != seen.end()) continue;
+      return_val += itr->first;
+      seen.insert(itr->first);
+      // for noutput > 1, add space separator
+      if (itr + 1 != strings.end()) return_val += " ";
+    }
+    return return_val;
+  } else {
+    return "Rewrite failed.";
+  }
+}
+
+void InverseTextNormalizer::GetGeneratedSymbolTable() {
+  const auto* symbolfst = grm_.GetFst("*StringFstSymbolTable");
+  if (symbolfst != nullptr) {
+    generated_symtab_ = std::shared_ptr<fst::SymbolTable>(
+        symbolfst->InputSymbols()->Copy());
+  }
+}
+
+void InverseTextNormalizer::FormatFst(fst::StdVectorFst *vfst) {
+  for (fst::StateIterator<fst::StdVectorFst> state_iter(*vfst);
+       !state_iter.Done(); state_iter.Next()) {
+    int state_id = state_iter.Value();
+    for (fst::MutableArcIterator<fst::StdVectorFst> arc_iter(vfst, state_id);
+         !arc_iter.Done(); arc_iter.Next()) {
+       const fst::StdArc arc = arc_iter.Value();
+       fst::StdArc new_arc(arc.ilabel & 0xff, arc.olabel & 0xff,
+                           arc.weight, arc.nextstate);
+       arc_iter.SetValue(new_arc);
+    }
+  }
+}
```

Comment:
The input and output of itn.fst are all bytes, not chars. For Chinese character (3 bytes), the highest bit of each byte may be 1 (i.e. 11100101), openfst will implicitly convert byte to int (note that int is the type of fst::Arc:iLabel), then we will get 11111111111111111111111111100101, this is a negative number and will fail the subsequent process.  Line 166 can make sure that all ilabels and labels of fst:arc are non-negtive, this is crucial if we want to reconstruct string from olabel sequence.